### PR TITLE
Support Azure Blob Storage for Cloud Storage Provider

### PR DESCRIPTION
**Pull request by yokawasa**
### Description

Support Azure Blob Storage as Cloud Storage Provider 
Relevant Issue: https://github.com/st-tech/gatling-operator/issues/38

## What's been changed / added?
- [Add Azure Blob Storage for Cloud Storage Provider and its test](https://github.com/st-tech/gatling-operator/pull/59/commits/c7146abd2a561a2d12898f4a069a5750c537ae68)
- [refactored csp so that init func is called in initialization](https://github.com/st-tech/gatling-operator/commit/da251eb545e9ae24f082ef41c38ba8c591a70779)
- [update CRD to support Azure CSP](https://github.com/st-tech/gatling-operator/commit/f65273ae1a80e2056d8aa546b22892eb0e06a868)
- [Update API Docs](https://github.com/st-tech/gatling-operator/commit/fa8449f45e9679aae07d9c178ff831ce1e3c73ef)
- [Add Azure configuration to sample gatling manifests](https://github.com/st-tech/gatling-operator/commit/92da305e13a67dd04404652370cef2c6799a614c)
- [Add Azure Blob Storage to README description](https://github.com/st-tech/gatling-operator/commit/576645be674793889f311482e1109a0925d7f3ae)

## How to configure Gatling CR to use Azure Blob Storage?

There are 2 ways to configure  Azure Blob Storage access authentication

### (1). Authenticating with Azure Blob Storage using AZUREBLOB_ACCOUNT and AZUREBLOB_KEY 

```yaml
provider: "azure"
bucket: "gatling-operator-reports"                   # Azure Blob Storage container name on which Gatlilng report files are stored
env:                                                 # Optional. Environment variables to be used for connecting to the cloud providers
      - name: AZUREBLOB_ACCOUNT                          # Azure Blob Storage Account Name
        value: xxxxxxxxxxxxxxx
      - name: AZUREBLOB_KEY                              # Azure Blob Access Key. Leave blank to use SAS URL
        valueFrom:
          secretKeyRef:
            name: azure-credentail-secrets
            key: AZUREBLOBSS_KEY
```

### (2).Authenticating with Azure Blob Storage using AZUREBLOB_ACCOUNT and AZUREBLOB_SAS_URL

```yaml
provider: "azure"
bucket: "gatling-operator-reports"                   # Azure Blob Storage container name on which Gatlilng report files are stored
env:                                                 # Optional. Environment variables to be used for connecting to the cloud providers
      - name: AZUREBLOB_ACCOUNT                          # Azure Blob Storage Account Name
        value: xxxxxxxxxxxxxxx
      - name: AZUREBLOB_KEY                              # Azure Blob Access Key. Leave blank to use SAS URL
        valueFrom:
          secretKeyRef:
            name: azure-credentail-secrets
            key: AZUREBLOBSS_KEY
```

Gatling Operator use [rclone](https://rclone.org/) to access any Cloud Storages, and to know more on Azure Blob Storage access using rclone, please see [Authenticating with Azure Blob Storage](https://rclone.org/azureblob/#authenticating-with-azure-blob-storage)

### Testing

1. prepare Azure blob storage account and container for testing (here, I used `gatlingreport001` and `gatling-operator-reports` respectively)

2. configure cloud storage provider to use azure blob storage in gatling CR to test authentication pattern ( 1 ) above

> config/samples/gatling-operator_v1alpha1_gatling01.yaml

```yaml
 cloudStorageSpec:
    provider: "azure"
    bucket: "gatling-operator-reports"
    env:                                             
      - name: AZUREBLOB_ACCOUNT
        value: gatlingreport001  
      - name: AZUREBLOB_KEY
        value: xxxxxxxxxxxxxxxxxxxxxxxxxxx==
```

apply the manifest and see the result

```

# apply the manifest
kustomize build config/samples  | kubectl apply -f -

# check the gatling cr status
kubectl get gatling gatling-sample01 -o jsonpath='{@.status}' |jqresult
```

See the report URL
```
kubectl get gatling gatling-sample01 -o jsonpath='{@.status}' |jq

{
  "reportCompleted": true,
  "reportStoragePath": "az:gatling-operator-reports/gatling-sample01/953431934",
  "reportUrl": "https://gatlingreport001.blob.core.windows.net/gatling-operator-reports/gatling-sample01/953431932/index.html",
  "reporterJobName": "gatling-sample01-reporter",
  "reporterStartTime": 1661659345,
  "runnerCompleted": true,
  "runnerJobName": "gatling-sample01-runner",
  "runnerStartTime": 1661658938,
  "succeeded": 3
}
```

made it access to reportURL (azure blob storage URL) like this:

Screen Shot 2022-08-28 at 13 05 45

3. prepare SAS URL for the azure blob storage to test authentication pattern ( 2 ) above

```yaml
 cloudStorageSpec:
   provider: "azure"
    bucket: "gatling-operator-reports" 
    env:
      - name: AZUREBLOB_SAS_URL
        value: "https://gatlingreport001.blob.core.windows.net/gatling-operator-reports?sp=rwl&st=2022-08-27T04:38:09Z&se=2022-09-01T12:38:09Z&spr=https&sv=2021-06-08&sr=c&sig=xGjILJf%2FJKZTPXXASIE5CAJa%2B6D%2BzO%2BJUWGthbjln5A%3D"
      - name: AZUREBLOB_ACCOUNT
        value: gatlingreport001
```

apply the manifest and see the result

```

# apply the manifest
kustomize build config/samples  | kubectl apply -f -

# check the gatling cr status
kubectl get gatling gatling-sample01 -o jsonpath='{@.status}' |jqresult
```

Likewise 2, made it access to reportURL (azure blob storage URL) !!

### Checklist

_Please check if applicable_

- [x] Tests have been added (if applicable, ie. when operator codes are added or modified)
- [x] Relevant docs have been added or modified (if applicable, ie. when new features are added or current features are modified) 
  - add API Docs
  - NOTE: relevant docs are not yet in  [User Guide](https://github.com/st-tech/gatling-operator/blob/main/docs/user-guide.md). It will be added in the next PR

Relevant issue #

https://github.com/st-tech/gatling-operator/issues/38

### CODE REVIEW

File: docs/api.md [hunk 1]
```diff
@@ -115,14 +115,14 @@ _Appears in:_
 | Field | Description |
 | --- | --- |
 | `startTime` _string_ | (Optional) Test Start time. |
-| `parallelism` _integer_ | (Optional) Number of pods running at the same time. Defaults to `1` (Minimum `1`) |
+| `parallelism` _integer_ | (Optional) Number of pods running at the same time. Defaults to `1` (Mininum `1`) |
```
Comment (ksudate) [inline docs/api.md:118]: Minimum?

Comment (yokawasa) [reply docs/api.md]: thanks! I fixed the typo https://github.com/st-tech/gatling-operator/pull/59/commits/9d7aef2e04c2ed3a7406d080f686e0024171e3a4

File: config/samples/gatling-operator_v1alpha1_gatling02.yaml [hunk 2]
```diff
@@ -37,6 +40,32 @@ spec:
     #      secretKeyRef:
     #        name: aws-credentail-secrets
     #        key: AWS_SECRET_ACCESS_KEY
+    #
+    #############################################################
+    #        Storage Provider - gcp (Google Cloud Storage)
+    #############################################################
+    # provider: "gcp"
+    # bucket: "gatling-operator-reports"                  # GCS bucket name on which Gatlilng report files are stored
+    # bucket: "zozo-mlops-dev-gatling-operator-reports"   # レポート用のGCSバケット zozo-mlops-dev(stg)-gatling-operator-reports を定義しているのでそちらを指定してください
```
Comment (akitok) [inline config/samples/gatling-operator_v1alpha1_gatling02.yaml:49]: I think this line is unnecessary.

File: pkg/cloudstorages/aws.go [hunk 3]
```diff
@@ -8,6 +8,8 @@ type AWSCloudStorageProvider struct {
 	providerName string
 }
 
+func (p *AWSCloudStorageProvider) init(args []EnvVars) { /* do nothinig */ }
```
Comment (akitok) [inline pkg/cloudstorages/aws.go:11]: nits: typo
```suggestion
func (p *AWSCloudStorageProvider) init(args []EnvVars) { /* do nothing */ }
```

File: pkg/cloudstorages/gcp.go [hunk 4]
```diff
@@ -1,11 +1,15 @@
...
-import "fmt"
+import (
+	"fmt"
+)
 
 type GCPCloudStorageProvider struct {
 	providerName string
 }
 
+func (p *GCPCloudStorageProvider) init(args []EnvVars) { /* do nothinig */ }
```
Comment (akitok) [inline pkg/cloudstorages/gcp.go:11]: nits: typo
```suggestion
func (p *GCPCloudStorageProvider) init(args []EnvVars) { /* do nothing */ }
```

File: config/samples/gatling-operator_v1alpha1_gatling01.yaml [hunk 5]
```diff
@@ -42,12 +47,39 @@ spec:
...
+    #
+    #############################################################
+    #        Storage Provider - gcp (Google Cloud Storage)
+    #############################################################
+    # provider: "gcp"
+    # bucket: "gatling-operator-reports"                  # GCS bucket name on which Gatlilng report files are stored
+    # bucket: "zozo-mlops-dev-gatling-operator-reports"   # レポート用のGCSバケット zozo-mlops-dev(stg)-gatling-operator-reports を定義しているのでそちらを指定してください
+    #
+    #############################################################
+    # .      Storage Provider - azure (Azure Blob Storage) ########
```
Comment (akitok) [inline config/samples/gatling-operator_v1alpha1_gatling01.yaml:59]: nits
```suggestion
    #       Storage Provider - azure (Azure Blob Storage)
```

File: config/samples/gatling-operator_v1alpha1_gatling02.yaml [hunk 6]
```diff
@@ -37,6 +40,32 @@ spec:
...
+    #
+    #############################################################
+    # .      Storage Provider - azure (Azure Blob Storage) ########
```
Comment (akitok) [inline config/samples/gatling-operator_v1alpha1_gatling02.yaml:52]: nits
```suggestion
    #       Storage Provider - azure (Azure Blob Storage)
```

File: config/samples/gatling-operator_v1alpha1_gatling01.yaml [hunk 7]
```diff
@@ -3,10 +3,12 @@ kind: Gatling
 metadata:
   name: gatling-sample01
 spec:
-  generateReport: false                                   # The flag of generating gatling report
+  generateReport: true                                   # The flag of generating gatling report
+  #generateReport: false                                   # The flag of generating gatling report
```
Comment (akitok) [inline config/samples/gatling-operator_v1alpha1_gatling01.yaml:7]: In this sample, I think `generateReport` is intended to be false.

File: config/samples/gatling-operator_v1alpha1_gatling01.yaml [hunk 8]
```diff
@@ -3,10 +3,12 @@ kind: Gatling
...
   generateLocalReport: false                              # The flag of generating gatling report for each pod
   notifyReport: false                                     # The flag of notifying gatling report
-  cleanupAfterJobDone: true                               #  The flag of cleaning up gatling jobs resources after the job done
+  cleanupAfterJobDone: false                               #  The flag of cleaning up gatling jobs resources after the job done
+  #cleanupAfterJobDone: true                               #  The flag of cleaning up gatling jobs resources after the job done
```
Comment (akitok) [inline config/samples/gatling-operator_v1alpha1_gatling01.yaml:11]: In this sample, I think `cleanupAfterJobDone` is intended to be true, is it not?

File: config/samples/gatling-operator_v1alpha1_gatling01.yaml [hunk 9]
```diff
@@ -42,12 +47,39 @@ spec:
     #      secretKeyRef:
     #        name: aws-credentail-secrets
     #        key: AWS_SECRET_ACCESS_KEY
+    #
+    #############################################################
+    #        Storage Provider - gcp (Google Cloud Storage)
+    #############################################################
+    # provider: "gcp"
+    # bucket: "gatling-operator-reports"                  # GCS bucket name on which Gatlilng report files are stored
+    # bucket: "zozo-mlops-dev-gatling-operator-reports"   # レポート用のGCSバケット zozo-mlops-dev(stg)-gatling-operator-reports を定義しているのでそちらを指定してください
```
Comment (akitok) [inline config/samples/gatling-operator_v1alpha1_gatling01.yaml:56]: nits: I think this line is unnecessary.

File: config/samples/gatling-operator_v1alpha1_gatling02.yaml (hunk 2 above)
Comment (akitok) [inline config/samples/gatling-operator_v1alpha1_gatling02.yaml:49]: nits: I think this line is unnecessary.

File: config/samples/gatling-operator_v1alpha1_gatling01.yaml [hunk 10]
```diff
@@ -42,12 +47,39 @@ spec:
...
+    #        name: azure-credentail-secrets
+    #        key: AZUREBLOB_SAS_URL
   notificationServiceSpec:
     provider: "slack"                                     # Notification provider name. Supported provider: "slack"
     secretName: "gatling-notification-slack-secrets"      # The name of secret in which all key/value sets needed for the notification are stored
   testScenarioSpec:
     # startTime: 2021-09-10 08:45:31                      # Optional. Start time of running test scenario in UTC. Format: %Y-%m-%d %H:%M:%S
-    parallelism: 3                                        # Optional. Default: 1. Number of pods running at any instant
+    parallelism: 1                                        # Optional. Default: 1. Number of pods running at any instant
+    #parallelism: 3                                        # Optional. Default: 1. Number of pods running at any instant
```
Comment (akitok) [inline config/samples/gatling-operator_v1alpha1_gatling01.yaml:82]: In this sample, I think `parallelism` is intended to be `3`, is it not?

File: config/samples/gatling-operator_v1alpha1_gatling01.yaml (hunk 7 above)
Comment (yokawasa) [reply config/samples/gatling-operator_v1alpha1_gatling01.yaml:7]: you're light!  it's just for testing

File: config/samples/gatling-operator_v1alpha1_gatling01.yaml (hunk 8 above)
Comment (yokawasa) [reply config/samples/gatling-operator_v1alpha1_gatling01.yaml]: you're light!  it's just for testing. https://github.com/st-tech/gatling-operator/pull/59/commits/c49c5091eca5c3ca5a0f5e77d17667e2c98acf21

File: config/samples/gatling-operator_v1alpha1_gatling02.yaml (hunk 2 above)
Comment (yokawasa) [reply config/samples/gatling-operator_v1alpha1_gatling02.yaml]: thanks. removed https://github.com/st-tech/gatling-operator/pull/59/commits/c49c5091eca5c3ca5a0f5e77d17667e2c98acf21

File: config/samples/gatling-operator_v1alpha1_gatling01.yaml (hunk 10 above)
Comment (yokawasa) [reply config/samples/gatling-operator_v1alpha1_gatling01.yaml]: you're light. it was just for testing https://github.com/st-tech/gatling-operator/pull/59/commits/c49c5091eca5c3ca5a0f5e77d17667e2c98acf21

File: config/samples/gatling-operator_v1alpha1_gatling01.yaml (hunk 9 above)
Comment (yokawasa) [reply config/samples/gatling-operator_v1alpha1_gatling01.yaml]: thanks. removed https://github.com/st-tech/gatling-operator/commit/c49c5091eca5c3ca5a0f5e77d17667e2c98acf21

**Review by ksudate — APPROVED** [review]
lgtm!

**Comment by yokawasa** [comment]
@ksudate thanks for the review and aproval

@akitok please let me know if you have any more feedback!

File: config/samples/gatling-operator_v1alpha1_gatling02.yaml [hunk 11]
```diff
@@ -37,6 +40,31 @@ spec:
...
+    #provider: "azure"
+    #bucket: "gatling-operator-reports"                   # Azure Blob Storage container name on which Gatlilng report files are stored
+    #env:                                                 # Optional. Environment variables to be used for connecting to the cloud providers
+    #  - name: AZUREBLOB_ACCOUNT                          # Azure Blob Storage Account Name
+    #    value: xxxxxxxxxxxxxxx
+    #  - name: AZUREBLOB_KEY                              # Azure Blob Access Key. Leave blank to use SAS URL
+    #    valueFrom:
+    #      secretKeyRef:
+    #        name: azure-credentail-secrets
+    #        key: AZUREBLOBSS_KEY
```
Comment (akitok) [inline config/samples/gatling-operator_v1alpha1_gatling02.yaml:62]: typo?
```suggestion
    #        key: AZUREBLOB_KEY
```

Comment (yokawasa) [reply config/samples/gatling-operator_v1alpha1_gatling02.yaml:62]: thanks! yes, typo 😄

**Review by akitok — APPROVED** [review]
Looks great to me!!

**Comment by yokawasa** [comment]
now i can merge the PR confidently with 2 of you guys approval 😄

**Pull request merged**